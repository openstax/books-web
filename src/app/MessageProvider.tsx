import { shouldPolyfill } from '@formatjs/intl-pluralrules/should-polyfill';
import React from 'react';
import { IntlProvider } from 'react-intl';
import enMessages from './messages/en';

// https://formatjs.io/docs/polyfills/intl-pluralrules/#dynamic-import--capability-detection
async function polyfill(locale: 'en') {
  if (shouldPolyfill()) {
    await import('@formatjs/intl-pluralrules/polyfill');
  }

  // boolean added by the polyfill
  if ((Intl.PluralRules as (typeof Intl.PluralRules & {polyfilled?: boolean})).polyfilled) {
    switch (locale) {
      case 'en':
        await import('@formatjs/intl-pluralrules/locale-data/en');
        break;
      default:
        throw new Error('unsupported locale in messageProvider');
    }
  }
}

polyfill('en');

// tslint:disable-next-line:variable-name
const MessageProvider: React.SFC<{onError?: () => void}> = (props) =>
  <IntlProvider onError={props.onError} locale='en' messages={enMessages}>{props.children}</IntlProvider>;

export default MessageProvider;
