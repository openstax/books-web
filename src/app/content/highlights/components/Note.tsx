import { HTMLTextAreaElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components/macro';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import { cardPadding, cardWidth } from '../constants';

interface Props {
  note: string;
  onChange: (note: string) => void;
  onFocus: () => void;
}

const noteMaxLength = 1000;

const width = cardWidth - cardPadding * 2;
// tslint:disable-next-line:variable-name
const TextArea = styled.textarea`
  display: block;
  min-height: 5.6rem;
  width: ${width}rem;
  max-height: 30rem;
  max-width: ${width}rem;
  min-width: ${width}rem;
  border: 1px solid ${theme.color.neutral.formBorder};
  padding: ${cardPadding}rem;
  ${textStyle}
  color: ${theme.color.text.label};
  font-size: 1.4rem;
  font-family: inherit;
  line-height: 2rem;
  font-weight: normal;
`;

// tslint:disable-next-line:variable-name
const Note = ({onChange, onFocus, note}: Props) => {
  const textArea = React.useRef<HTMLTextAreaElement>(null);

  const setTextAreaHeight = () => {
    const element = textArea.current;
    if (!element) {
      return;
    }

    if (element.scrollHeight > element.offsetHeight) {
      element.style.height = `${element.scrollHeight + 5}px`;
    }
  };

  React.useEffect(setTextAreaHeight, [note]);

  return <FormattedMessage id='i18n:highlighting:card:placeholder'>
    {(msg: Element | string) => <TextArea
        ref={textArea}
        value={note}
        onFocus={onFocus}
        maxLength={noteMaxLength}
        onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
          onChange(e.target.value);
        }}
        placeholder={msg}
      />
    }
  </FormattedMessage>;
};

export default Note;
