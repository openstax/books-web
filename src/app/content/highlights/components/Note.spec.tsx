import React from 'react';
import renderer from 'react-test-renderer';
import { renderToDom } from '../../../../test/reactutils';
import MessageProvider from '../../../MessageProvider';
import Note from './Note';

describe('Note', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<MessageProvider>
      <Note note='' onChange={() => null} onFocus={() => null} />
    </MessageProvider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onChange', () => {
    const onChange = jest.fn();
    const component = renderer.create(<MessageProvider>
      <Note note='' onChange={onChange} onFocus={() => null} />
    </MessageProvider>);

    const textarea = component.root.findByType('textarea');

    renderer.act(() => {
      textarea.props.onChange({
        target: {
          value: 'asdf',
        },
      });
    });

    expect(onChange).toHaveBeenCalledWith('asdf');
  });

  it('resizes on update when necessary', () => {
    const {node, root} = renderToDom(<MessageProvider>
      <Note note='' onChange={() => null} onFocus={() => null} />
    </MessageProvider>);

    Object.defineProperty(node, 'scrollHeight', { value: 100 });
    Object.defineProperty(node, 'offsetHeight', { value: 50 });

    renderToDom(<MessageProvider>
      <Note note='asdf' onChange={() => null} onFocus={() => null} />
    </MessageProvider>, root);

    expect(node.style.height).toEqual('105px');
  });

  it('doesn\'t resize on update when unneccessary', () => {
    const {node, root} = renderToDom(<MessageProvider>
      <Note note='' onChange={() => null} onFocus={() => null} />
    </MessageProvider>);

    Object.defineProperty(node, 'scrollHeight', { value: 50 });
    Object.defineProperty(node, 'offsetHeight', { value: 50 });

    renderToDom(<MessageProvider>
      <Note note='asdf' onChange={() => null} onFocus={() => null} />
    </MessageProvider>, root);

    expect(node.style.height).toEqual('');
  });
});
