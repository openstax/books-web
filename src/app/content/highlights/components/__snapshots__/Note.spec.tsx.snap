// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note matches snapshot 1`] = `
<textarea
  className="Note__TextArea-sc-1lk7g7-0 gPZklF"
  maxLength={1000}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="Add a note..."
  value=""
/>
`;
